File: hse/BPM161/abramov/2/LockFreeSet/src/main/java/ru/spbhse/karvozavr/lockfreeset/LockFreeSet.java
package ru.spbhse.karvozavr.lockfreeset;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.concurrent.atomic.AtomicStampedReference;

/**
 * Lock-free set implementation using linked list.
 *
 * @param <T> Key type
 */
public class LockFreeSet<T extends Comparable<T>> implements LockFreeSetInterface<T> {

    private static final int REMOVED = -1;
    private static final int INITIAL = 0;

    /**
     * List node.
     */
    private class Node {

        final T value;

        AtomicStampedReference<Node> next = new AtomicStampedReference<>(null, INITIAL);

        Node(T value) {
            this.value = value;
        }
    }

    /**
     * Stamped reference to the head of the list.
     * Head is fictive Node with null value.
     */
    private final AtomicStampedReference<Node> head = new AtomicStampedReference<>(new Node(null), INITIAL);

    @Override
    public boolean add(@NotNull T value) {
        Node node = new Node(value);

        while (true) {
            Pair<Node> prevAndCurrent = find(value);
            Node prev = prevAndCurrent.first;
            Node current = prevAndCurrent.second;

            if (current != null) {
                return false;
            } else {
                var stamp = prev.next.getStamp();
                if (stamp != REMOVED) {
                    if (prev.next.compareAndSet(null, node, stamp, stamp + 1)) {
                        return true;
                    }
                } else {
                    fixRemove();
                }
            }
        }
    }

    @Override
    public boolean remove(@NotNull T value) {
        while (true) {
            Pair<Node> prevAndCurrent = find(value);
            Node prev = prevAndCurrent.first;
            Node current = prevAndCurrent.second;

            if (current != null) {
                Node ref = current.next.getReference();
                var stamp = current.next.getStamp();
                if (stamp != REMOVED && current.next.compareAndSet(ref, ref, stamp, REMOVED)) {
                    physicallyRemove(prev, current);
                    return true;
                }
            } else {
                return false;
            }
        }
    }

    @Override
    public boolean contains(T value) {
        if (value == null) {
            return false;
        }

        Node node = find(value).second;

        return node != null;
    }

    @Override
    public boolean isEmpty() {
        return !iterator().hasNext();
    }

    @Override
    public Iterator<T> iterator() {
        return getSnapshot().iterator();
    }

    private Pair<Node> find(@NotNull T value) {
        Node prev, node;
        for (prev = head.getReference(), node = prev.next.getReference();
             node != null; prev = node, node = node.next.getReference()) {
            if (!isRemoved(node) && value.equals(node.value)) {
                return new Pair<>(prev, node);
            }
        }

        return new Pair<>(prev, null);
    }

    private boolean isRemoved(@NotNull Node node) {
        return node.next.getStamp() == REMOVED;
    }

    private void fixRemove() {
        Node prev, node;
        for (prev = head.getReference(), node = prev.next.getReference();
             node != null; prev = node, node = node.next.getReference()) {
            if (isRemoved(node)) {
                physicallyRemove(prev, node);
            }
        }
    }

    private void physicallyRemove(@NotNull Node prev, @Nullable Node node) {
        Node next = node;
        while (next != null && isRemoved(next)) {
            next = next.next.getReference();
        }

        var stamp = prev.next.getStamp();
        if (stamp != REMOVED) {
            prev.next.compareAndSet(node, next, stamp, stamp + 1);
        }
    }

    private List<T> getSnapshot() {
        List<T> elements;
        List<Integer> versions;

        attempt:
        while (true) {
            elements = new ArrayList<>();
            versions = new ArrayList<>();

            Node node = head.getReference();
            versions.add(node.next.getStamp());
            for (node = node.next.getReference(); node != null; node = node.next.getReference()) {
                if (!isRemoved(node)) {
                    elements.add(node.value);
                    versions.add(node.next.getStamp());
                }
            }

            var elemIter = elements.iterator();
            var versionIter = versions.iterator();

            node = head.getReference();
            if (node.next.getStamp() != versionIter.next()) {
                continue attempt;
            }
            for (node = head.getReference().next.getReference(); node != null; node = node.next.getReference()) {
                if (!isRemoved(node)) {
                    if (!elemIter.hasNext() || elemIter.next() != node.value || versionIter.next() != node.next.getStamp()) {
                        continue attempt;
                    }
                }
            }

            break attempt;
        }

        return elements;
    }
}

File: hse/BPM161/egorov/lab_2/src/main/kotlin/ru/hse/egorov/LockFreeSet.kt
package ru.hse.egorov

import java.util.concurrent.atomic.AtomicMarkableReference

class LockFreeSet<T : Comparable<T>> : LockFreeSetInterface<T> {
    private val head = AtomicMarkableReference(
        Node(
            null,
            AtomicMarkableReference<Node<T>>(null, false)
        ), true
    )


    override val isEmpty: Boolean
        get() {
            var firstSnapshot = takeSnapshot()
            var secondSnapshot = takeSnapshot()
            while (firstSnapshot != secondSnapshot) {
                firstSnapshot = takeSnapshot()
                secondSnapshot = takeSnapshot()
            }
            return firstSnapshot.isEmpty()
        }

    override fun add(value: T): Boolean {
        val node = Node(value, AtomicMarkableReference<Node<T>>(null, false))

        while (true) {
            val (prev, current) = find(value)

            if (current != null) {
                return false
            } else {
                if (prev.next.compareAndSet(null, node, false, false)) {
                    return true
                }
            }
        }
    }

    override fun remove(value: T): Boolean {
        while (true) {
            val (prev, current) = find(value)

            if (current == null)
                return false

            val ref = current.next.reference
            if (current.next.compareAndSet(ref, ref, false, true)) {
                physicallyRemove(prev, current)
                return true
            }
        }
    }


    override fun contains(value: T): Boolean {
        var cur = head.reference.next.reference
        while (cur != null) {
            if (cur.value == value)
                return !isRemoved(cur)
            cur = cur.next.reference
        }
        return false
    }

    override fun iterator(): Iterator<T> {
        var firstSnapshot = takeSnapshot()
        var secondSnapshot = takeSnapshot()
        while (firstSnapshot != secondSnapshot) {
            firstSnapshot = takeSnapshot()
            secondSnapshot = takeSnapshot()
        }
        return firstSnapshot.map { ref -> ref.value!! }.iterator()
    }

    private fun takeSnapshot(): List<Node<T>> {
        var cur = head.reference.next.reference
        val set = mutableListOf<Node<T>>()

        while (cur != null) {
            set.add(cur)
            cur = cur.next.reference
        }

        return set.filter { ref -> !isRemoved(ref) }
    }

    private fun find(value: T): Pair<Node<T>, Node<T>?> {
        var prev: Node<T>
        var node: Node<T>?
        prev = head.reference
        node = prev.next.reference
        while (node != null) {
            if (!isRemoved(node) && value == node.value) {
                return Pair(prev, node)
            }
            prev = node
            node = node.next.reference
        }

        return Pair<Node<T>, Node<T>?>(prev, node)
    }

    private fun isRemoved(node: Node<T>): Boolean {
        return node.next.isMarked
    }

    private fun physicallyRemove(prev: Node<T>, node: Node<T>) {
        var next: Node<T>? = node
        do {
            next = next!!.next.reference
        } while (next != null && isRemoved(next))

        prev.next.compareAndSet(node, next, false, false)
    }

    companion object {
        private data class Node<T>(val value: T?, val next: AtomicMarkableReference<Node<T>>)
    }
}